
# What is ML & sklearn ?
***
머신러닝(Machine Learning)이란 무엇일까요? 학부생인 저의 지식과 경험에 비추어 보면 머신러닝은 간단하게 알고리즘 정도로 설명이 가능할 것 같습니다. 조금더 보태면 데이터로 어떠한 현상을 예측하고자 하는 기법을 일컫는다고 생각하면 될 것 입니다.

현재 머신러닝 기법이 사회 전반적인 분야는 물론 개인의 취향 및 건강 관련 서비스에도 통용이 되고 있습니다. 데이터가 내포하고 있는 패턴을 학습시켜서 새로운 형태의 데이터가 주어졌을 때 해당 데이터가 의미하는 바를 정확하게 잡아내기 위한 것이 중요한데요. 스팸메일을 필터링하는 모델을 만든다고 가정했을 때 특정 단어가 해당 메일에 등장했다는 이유만으로 메일을 스팸메일로 분류한다면 또 다른 문제점을 낳을 수 있다고 생각합니다.

이러한 이유들로 날이 갈수록 머신러닝의 기법들의 수준도 점점 고도화 되어가고 있고 머신러닝이 더 상용화 되고 보편화 돼가고 있습니다.

그렇다면 머신러닝의 종류에는 어떤 것들이 있을까요? 머신러닝에는 크게 지도학습(supervised learning)과 비지도학습(unsupervised leaerning)이 있습니다.

대표적인 지도학습의 종류는 아래와 같습니다.
- 분류
- 회귀
- NLP(자연어 처리)

또한 대표적인 비지도학습의 종류는 아래와 같습니다.

- 군집화
- 차원축소
- 강화학습

지도학습과 비지도학습의 차이점은 이후 다른 챕터에서 구체적으로 설명을 드리도록 하겠습니다. 머신러닝에서 가장 중요한 요소는 무엇일까요? 뛰어난 알고리즘과 많은 데이터 등 한 요소만을 특정해서 중요하다고 할 수가 없습니다. 하지만 빅데이터가 급부상하는 현재 데이터는 이곳 저곳에 정말 무수히 많습니다. 그렇다면 이런 데이터들을 무작정 가져다 사용한다고 좋은 결과가 나오고 만약 좋은 결과가 나온다고 해도 그 결과를 신뢰할 수 있을까요?

편중된 1억개의 데이터 보다 대표성이 높은 100개의 데이터가 더 좋은 성능과 높은 신뢰도를 가져다 줄 수 있습니다. 그렇기 때문에 좋은 품질의 데이터를 갖추었을 때 머신러닝의 수행 결과도 좋다고 할 수 있겠습니다. 즉 좋은 품질의 데이터와 해당 데이터에 맞는 적절한 알고리즘 등 여러가지 요소가 조화를 이루었을 때 우리는 비로소 뛰어나고 신뢰할만한 머신러닝의 수행 결과를 얻을 수 있습니다.

자, 그럼 이제 실제로 머신러닝이 어떤 과정을 통해서 수행이 되는지 살펴보도록 하겠습니다. 우선 '사이킷런(scikit-learn)은 파이썬에서 사용되는 ML 라이브러리 중 가장 많이 사용되는 라이브러리입니다. 특히 정형 데이터를 활용한 머신러닝에서 우수한 성능을 보여줍니다. 주로 코드를 활용해서 설명이 진행되겠고 각 코드에는 주석을 달아놔서 쉽게 이해할 수 있도록 하겠습니다.

먼저 머신러닝에서 아주 대표적 데이터셋인 붓꽃 데이터 세트 일명 iris를 활용해서 분류 모델을 만들어보도록 하겠습니다. 분류는 앞서 말씀드렸듯이 지도학습의 대표적 기법 중 하나인데요. 지도학습은 정답(label 혹은 target)이 있는 데이터로 모델을 학습한 후 정답이 주어지지 않은 데이터에 대해 정답을 예측하는 방법입니다. 수능 성적 배치표를 가지고 내가 어느 대학에 갈 수 있을지 판단해보는것과 같다고 보시면 될 것 같습니다.

In [2]:
# 사이킷런이 설치가 안되어 있다면 아래 코드를 실행해 주세요.
!pip install scikit-learn # jupyternotebook 상에서 실행

#### 사이킷런의 다양한 패키지들에 대해 보고싶은 분들은 scikit-learn 공식 문서를 참고하시길 바랍니다.


간단한 의사결정트리(DecisionTreeClassifier) 알고리즘을 활용하여 붓꽃 분류 문제를 진행해보도록 하겠습니다.

우선 필요한 패키지들을 호출합니다.

In [3]:
from sklearn.datasets import load_iris # 붓꽃 데이터셋 호출
from sklearn.tree import DecisionTreeClassifier # 의사결정트리 호출
from sklearn.model_selection import train_test_split # 학습용 데이터와 테스트용 데이터로 분할 함수
from sklearn.metrics import accuracy_score # 평가지표에서 정확도 호출
import pandas as pd # 데이터 전처리용 패키지

In [4]:
iris = load_iris() # 붓꽃 데이터 data 변수에 저장

data = iris.data # 붓꽃 데이터의 정답을 제외한 데이터
target = iris.target # 붓꽃 데이터의 정답 데이터

In [5]:
data[:3] # 3번째 데이터까지만 호출


array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2]])

In [7]:
target[:3] # 3번째 정답 데이터까지만 호출

array([0, 0, 0])

In [8]:
df = pd.DataFrame(data, columns = iris.feature_names) # 컬럼명과 함께 보기 위해 데이터 프레임으로 변환
df['정답'] = target # 데이터의 정답 컬럼에 iris target값 대입
df.head() # 최종 데이터의 5행까지 출력

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   정답  
0   0  
1   0  
2   0  
3   0  
4   0

In [11]:
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size = .2, random_state = 203)
# 데이터를 학습용(train)과 검증용(test)으로 나눕니다.

In [12]:
dt = DecisionTreeClassifier(random_state = 333) # 의사결정나무를 생성합니다.

In [13]:
dt.fit(X_train, y_train) # 학습용 데이터를 모델로 학습합니다.

DecisionTreeClassifier(random_state=333)

In [14]:
pred = dt.predict(X_test) # 학습시킨 모델로 검증용 데이터를 예측합니다.

In [15]:
# accuracy_score(예측값, 실제값)을 호출하면 정확도를 계산합니다.
print('정확도는 {}'.format(accuracy_score(pred, y_test)))

정확도는 0.9333333333333333


지금까지 아주 간단한 분류 문제를 해결해보았습니다. 93.3%라는 높은 정확도를 가지는 모델도 만들었습니다. 일반적인 머신러닝 모델 알고리즘 적용은 데이터를 학습한 후 예측하고자 하는 데이터를 예측하는 과정으로 진행됩니다.


사이킷런에는 기본적으로 내장되어 있는 데이터들이 있습니다. 회귀와 분류용 데이터셋들이 있으며 종류는 아래와 같습니다.

- load_boston() : 보스턴 주택 가격 데이터셋
- load_breast_cancer() : 위스콘신 유방암 악성/음성 데이터셋
- load_diabetes() : 당뇨 데이터셋
- load_digits() : 0에서 9까지 숫자 이미지 데이터셋
- load_iris() : 붓꽃 데이터셋

따라서 이후에 다양한 실습을 하고 싶으신 분들은 위의 데이터셋을 활용하시면 될 것 같습니다.

사이킷런에는 데이터 전처리, 모델 검증 및 파라미터 튜닝 등 머신러닝 문제를 해결하기 위한 다양한 패키지들이 존재합니다. 따라서 본인이 풀고자 하는 문제에 대한 적절한 접근법을 활용하여 해당 문제를 해결하면 되겠습니다.

지금까지 머신러닝과 사이킷런에 대해서 매우 간단하고 얕게 살펴보았습니다. 이후에는 다양한 코드들과 함께 보다 자세하게 내용을 다룰 예정입니다. 다음 챕터에서는 분류(classification)에 대해 설명을 드리도록 하겠습니다.